lib/Sema/CSDiag.cpp

```diff
@@ -7200,18 +7200,19 @@ bool FailureDiagnosis::visitBindOptionalExpr(BindOptionalExpr *BOE) {
 }
 
 bool FailureDiagnosis::visitIfExpr(IfExpr *IE) {
-  auto typeCheckClauseExpr = [&](Expr *clause) -> Expr * {
-    return typeCheckChildIndependently(clause, Type(), CTP_Unused, TCCOptions(),
+  auto typeCheckClauseExpr = [&](Expr *clause, ContextualTypePurpose convertTypePurpose) -> Expr * {
+      // Provide proper contextual type when type conversion is specified.
+      auto type = convertTypePurpose == CTP_Unused ? Type() : CS.getContextualType();
+      return typeCheckChildIndependently(clause, type, convertTypePurpose, TCCOptions(),
```

Comment:
Drive-by nit: please adhere to 80-column lines and indent two spaces in this file.